File: src/java/com/codecool/gameoflife/view/SetupView.java
package com.codecool.gameoflife.view;

import com.codecool.gameoflife.controller.GameController;
import com.codecool.gameoflife.model.RuleSet;
import javafx.beans.value.ChangeListener;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.geometry.Insets;
import javafx.scene.control.*;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;

public class SetupView extends HBox {

    private VBox gameSettings;
    private VBox rulesSettings;
    private Button startButton;
    private Button randomizeButton;
    private Button clearButton;
    private Slider speedSlider;
    private CheckBox borderLessModeToggle;
    private CheckBox gridVisibleToggle;
    private ToggleGroup rulesToggle;

    SetupView() {
        super(50);
        this.gameSettings = new VBox(10);
        this.rulesSettings = new VBox(10);
        this.setPadding(new Insets(25));
        this.startButton = new Button("Start");
        this.randomizeButton = new Button("Randomize grid");
        this.clearButton = new Button("Clear");
        Label speedLabel = new Label("Speed");
        this.speedSlider = new Slider(0, 400, 200);
        this.speedSlider.setShowTickLabels(true);
        this.borderLessModeToggle = new CheckBox("Borderless mode");
        this.gridVisibleToggle = new CheckBox("Grid visible");
        this.rulesToggle = new ToggleGroup();
        Label rulesLabel = new Label("Rules modifications:");
        this.gameSettings.getChildren().addAll(startButton, randomizeButton, clearButton, speedLabel,
                speedSlider, borderLessModeToggle, gridVisibleToggle);
        this.rulesSettings.getChildren().add(rulesLabel);
        this.getChildren().addAll(gameSettings, rulesSettings);
    }

    void setupRulesToggles(GameController gameController) {
        for (RuleSet ruleSet : RuleSet.values()) {
            RadioButton button = new RadioButton(ruleSet.toString());
            button.setToggleGroup(this.rulesToggle);
            button.setUserData(ruleSet);
            button.setOnMouseClicked(event -> gameController.setRules((RuleSet) button.getUserData()));
            this.rulesSettings.getChildren().add(button);
        }
    }

    void setupSlider(ChangeListener<Number> numberChangeListener) {
        this.speedSlider.valueProperty().addListener(numberChangeListener);
    }

    void setupStartButton(EventHandler<ActionEvent> handler, String label) {
       this.startButton.setText(label);
       this.startButton.setOnAction(handler);
    }

    void setupRandomizeButton(EventHandler<ActionEvent> handler) {
        this.randomizeButton.setOnAction(handler);
    }

    void setupClearButton(EventHandler<ActionEvent> handler) {
        this.clearButton.setOnAction(handler);
    }

    void setupBorderLessToggle(ChangeListener<Boolean> changeListener) {
        this.borderLessModeToggle.selectedProperty().addListener(changeListener);
    }

    void setupGridVisibleToggle(ChangeListener<Boolean> changeListener) {
        this.gridVisibleToggle.selectedProperty().addListener(changeListener);
    }
}

File: src/kotlin/com/codecool/gameoflife/view/SetupView.kt
package com.codecool.gameoflife.view

import com.codecool.gameoflife.controller.GameController
import com.codecool.gameoflife.model.RuleSet
import javafx.beans.value.ChangeListener
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.scene.control.*
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.scene.text.Text

class SetupView : HBox(50.0) {

    private val rulesSettings = VBox(10.0)
    private val startButton = Button("Start")
    private val randomizeButton = Button("Randomize grid")
    private val clearButton = Button("Clear")
    private val speedSlider = Slider(0.0, 400.0, 200.0)
    private val borderLessModeToggle = CheckBox("Borderless mode")
    private val gridVisibleToggle = CheckBox("Grid visible")
    private val rulesToggle = ToggleGroup()
    private val validationResult: Text = Text()

    init {
        this.padding = Insets(25.0)
        val speedLabel = Label("Speed")
        speedSlider.isShowTickLabels = true
        val gameSettings = VBox(10.0).apply {
            children.addAll(
                    startButton,
                    randomizeButton,
                    clearButton,
                    speedLabel,
                    speedSlider,
                    borderLessModeToggle,
                    gridVisibleToggle
            )
        }
        children.addAll(gameSettings, rulesSettings)
    }

    fun createRuleButtons(gameController: GameController) {
        rulesSettings.children.apply {
            setAll(Label("Rules modifications:"))
            RuleSet.ruleSets.values.forEach { rs ->
                add(RadioButton(rs.name).apply {
                    toggleGroup = rulesToggle
                    userData = rs
                    setOnMouseClicked { gameController.setRules(this.userData as RuleSet) }
                })
            }
            add(createCustomRulesForm(gameController))
            add(validationResult)
        }
    }

    private fun createCustomRulesForm(controller: GameController): HBox {
        val form = HBox(5.0)
        val aliveRules = TextField()
        val deadRules = TextField()
        aliveRules.prefColumnCount = 5
        deadRules.prefColumnCount = 5
        val addRules = Button("Add")
        addRules.setOnAction {
            val success = controller.addRules("Custom", aliveRules.text, deadRules.text)
            if (success) {
                validationResult.text = "Added new custom rules."
            } else {
                validationResult.text = "Wrong rules format"
            }
        }
        form.children.addAll(Label("Custom:"), aliveRules,
                Label("/"), deadRules, addRules)
        return form
    }

    fun setupSlider(numberChangeListener: ChangeListener<Number>) {
        speedSlider.valueProperty().addListener(numberChangeListener)
    }

    fun setupStartButton(handler: EventHandler<ActionEvent>, label: String) {
        startButton.text = label
        startButton.onAction = handler
    }

    fun setupRandomizeButton(handler: EventHandler<ActionEvent>) {
        randomizeButton.onAction = handler
    }

    fun setupClearButton(handler: EventHandler<ActionEvent>) {
        clearButton.onAction = handler
    }

    fun setupBorderLessToggle(changeListener: ChangeListener<Boolean>) {
        borderLessModeToggle.selectedProperty().addListener(changeListener)
    }

    fun setupGridVisibleToggle(changeListener: ChangeListener<Boolean>) {
        gridVisibleToggle.selectedProperty().addListener(changeListener)
    }
}
